File: server/services/trialSources/clinicalTrialsGov.ts
/**
 * ClinicalTrials.gov API Connector
 * ================================
 * Fetches clinical trial data from the ClinicalTrials.gov API v2
 * https://clinicaltrials.gov/data-api/api
 */

import { db } from "../../db";
import { clinicalTrials, dataSourceStatus } from "@shared/schema";
import { eq } from "drizzle-orm";

const API_BASE_URL = "https://clinicaltrials.gov/api/v2";
const SOURCE_NAME = "clinicaltrials_gov";

interface CTGStudy {
  protocolSection: {
    identificationModule: {
      nctId: string;
      briefTitle: string;
      officialTitle?: string;
      secondaryIdInfos?: Array<{ id: string; type?: string }>;
    };
    statusModule: {
      overallStatus: string;
      startDateStruct?: { date: string };
      completionDateStruct?: { date: string };
      lastUpdateSubmitDate?: string;
    };
    descriptionModule?: {
      briefSummary?: string;
      detailedDescription?: string;
    };
    conditionsModule?: {
      conditions?: string[];
    };
    designModule?: {
      studyType?: string;
      phases?: string[];
    };
    armsInterventionsModule?: {
      interventions?: Array<{
        type: string;
        name: string;
      }>;
    };
    eligibilityModule?: {
      eligibilityCriteria?: string;
      sex?: string;
      minimumAge?: string;
      maximumAge?: string;
      healthyVolunteers?: boolean;
    };
    contactsLocationsModule?: {
      locations?: Array<{
        facility?: string;
        city?: string;
        country?: string;
        status?: string;
      }>;
      centralContacts?: Array<{
        name?: string;
        email?: string;
        phone?: string;
        role?: string;
      }>;
    };
    sponsorCollaboratorsModule?: {
      leadSponsor?: {
        name?: string;
        class?: string;
      };
    };
  };
}

interface CTGSearchResponse {
  studies: CTGStudy[];
  nextPageToken?: string;
  totalCount?: number;
}

function getCountryCode(country: string): string {
  const countryMap: Record<string, string> = {
    "United States": "US",
    "Germany": "DE",
    "France": "FR",
    "United Kingdom": "GB",
    "Italy": "IT",
    "Spain": "ES",
    "Netherlands": "NL",
    "Belgium": "BE",
    "Switzerland": "CH",
    "Austria": "AT",
    "Poland": "PL",
    "Georgia": "GE",
    "Armenia": "AM",
    "Azerbaijan": "AZ",
    "Turkey": "TR",
    "Israel": "IL",
    "Canada": "CA",
    "Australia": "AU",
    "Japan": "JP",
    "China": "CN",
    "South Korea": "KR",
    "India": "IN",
  };
  return countryMap[country] || country.substring(0, 2).toUpperCase();
}

export async function searchTrials(
  query: string,
  options: {
    pageSize?: number;
    pageToken?: string;
    status?: string[];
    phase?: string[];
    ageRange?: { min?: string; max?: string };
    countries?: string[];
  } = {}
): Promise<{ trials: CTGStudy[]; nextPageToken?: string; totalCount: number }> {
  const params = new URLSearchParams({
    "query.term": query,
    pageSize: String(options.pageSize || 20),
    format: "json",
  });

  if (options.pageToken) {
    params.append("pageToken", options.pageToken);
  }

  if (options.status?.length) {
    params.append("filter.overallStatus", options.status.join(","));
  }

  const url = `${API_BASE_URL}/studies?${params.toString()}`;
  console.log(`[ClinicalTrials.gov] Searching: ${url}`);

  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`API error: ${response.status} ${response.statusText}`);
    }

    const data = (await response.json()) as CTGSearchResponse;
    
    return {
      trials: data.studies || [],
      nextPageToken: data.nextPageToken,
      totalCount: data.totalCount || 0,
    };
  } catch (error) {
    console.error("[ClinicalTrials.gov] Search error:", error);
    throw error;
  }
}

export async function getTrialByNCT(nctId: string): Promise<CTGStudy | null> {
  const url = `${API_BASE_URL}/studies/${nctId}?format=json`;
  
  try {
    const response = await fetch(url);
    if (!response.ok) {
      if (response.status === 404) return null;
      throw new Error(`API error: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error(`[ClinicalTrials.gov] Error fetching ${nctId}:`, error);
    return null;
  }
}

export function transformToDBFormat(study: CTGStudy) {
  const proto = study.protocolSection;
  const id = proto.identificationModule;
  const status = proto.statusModule;
  const desc = proto.descriptionModule;
  const cond = proto.conditionsModule;
  const design = proto.designModule;
  const arms = proto.armsInterventionsModule;
  const elig = proto.eligibilityModule;
  const contact = proto.contactsLocationsModule;
  const sponsor = proto.sponsorCollaboratorsModule;

  const eudraCTInfo = id.secondaryIdInfos?.find(
    (s) => s.type === "EudraCT Number"
  );

  return {
    nctNumber: id.nctId,
    eudraCTNumber: eudraCTInfo?.id || null,
    titleEn: id.officialTitle || id.briefTitle,
    briefSummaryEn: desc?.briefSummary || null,
    detailedDescriptionEn: desc?.detailedDescription || null,
    status: status.overallStatus,
    phase: design?.phases?.join(", ") || null,
    studyType: design?.studyType || null,
    eligibilityCriteriaEn: elig?.eligibilityCriteria || null,
    minAge: elig?.minimumAge || null,
    maxAge: elig?.maximumAge || null,
    gender: elig?.sex || "All",
    healthyVolunteers: elig?.healthyVolunteers || false,
    conditions: cond?.conditions || [],
    interventions:
      arms?.interventions?.map((i) => ({ type: i.type, name: i.name })) || [],
    locations:
      contact?.locations?.map((l) => ({
        facility: l.facility || "",
        city: l.city || "",
        country: l.country || "",
        countryCode: getCountryCode(l.country || ""),
        status: l.status,
      })) || [],
    contacts:
      contact?.centralContacts?.map((c) => ({
        name: c.name || "",
        email: c.email,
        phone: c.phone,
        role: c.role,
      })) || [],
    sponsorName: sponsor?.leadSponsor?.name || null,
    sponsorType: sponsor?.leadSponsor?.class || null,
    startDate: status.startDateStruct?.date || null,
    completionDate: status.completionDateStruct?.date || null,
    lastUpdateDate: status.lastUpdateSubmitDate
      ? new Date(status.lastUpdateSubmitDate)
      : null,
    sources: [SOURCE_NAME],
  };
}

export async function syncTrialsForCondition(
  condition: string,
  maxPages: number = 5
): Promise<{ added: number; updated: number; errors: number }> {
  console.log(`[ClinicalTrials.gov] Syncing trials for: ${condition}`);
  
  let added = 0;
  let updated = 0;
  let errors = 0;
  let pageToken: string | undefined;

  try {
    for (let page = 0; page < maxPages; page++) {
      const result = await searchTrials(condition, {
        pageSize: 50,
        pageToken,
        status: ["RECRUITING", "ENROLLING_BY_INVITATION", "ACTIVE_NOT_RECRUITING"],
      });

      for (const study of result.trials) {
        try {
          const trialData = transformToDBFormat(study);
          
          const existing = await db
            .select()
            .from(clinicalTrials)
            .where(eq(clinicalTrials.nctNumber, trialData.nctNumber!))
            .limit(1);

          if (existing.length > 0) {
            await db
              .update(clinicalTrials)
              .set({ ...trialData, updatedAt: new Date() })
              .where(eq(clinicalTrials.nctNumber, trialData.nctNumber!));
            updated++;
          } else {
            await db.insert(clinicalTrials).values(trialData as any);
            added++;
          }
        } catch (err) {
          console.error(`[ClinicalTrials.gov] Error saving trial:`, err);
          errors++;
        }
      }

      if (!result.nextPageToken) break;
      pageToken = result.nextPageToken;

      await new Promise((r) => setTimeout(r, 500));
    }

    await updateSourceStatus("success", added + updated);
    console.log(`[ClinicalTrials.gov] Sync complete: ${added} added, ${updated} updated, ${errors} errors`);

  } catch (error) {
    console.error("[ClinicalTrials.gov] Sync failed:", error);
    await updateSourceStatus("failed", 0, String(error));
  }

  return { added, updated, errors };
}

async function updateSourceStatus(
  status: string,
  recordsCount: number,
  errorMessage?: string
) {
  const existing = await db
    .select()
    .from(dataSourceStatus)
    .where(eq(dataSourceStatus.sourceName, SOURCE_NAME))
    .limit(1);

  if (existing.length > 0) {
    await db
      .update(dataSourceStatus)
      .set({
        lastSyncAt: new Date(),
        lastSyncStatus: status,
        recordsCount,
        errorMessage: errorMessage || null,
      })
      .where(eq(dataSourceStatus.sourceName, SOURCE_NAME));
  } else {
    await db.insert(dataSourceStatus).values({
      sourceName: SOURCE_NAME,
      sourceDisplayName: "ClinicalTrials.gov",
      lastSyncAt: new Date(),
      lastSyncStatus: status,
      recordsCount,
      isActive: true,
      syncIntervalHours: 24,
    });
  }
}

File: backend/scrapers/clinicaltrials_gov.py
"""ClinicalTrials.gov API scraper."""

from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta

from .base_scraper import BaseScraper


class ClinicalTrialsGovScraper(BaseScraper):
    """Scraper for ClinicalTrials.gov API v2."""

    BASE_URL = "https://clinicaltrials.gov/api/v2/studies"

    def __init__(self):
        super().__init__()
        self.rate_limit = 3.0  # 3 requests per second

    @property
    def source_name(self) -> str:
        return "clinicaltrials.gov"

    async def search(
        self,
        condition: str,
        status: List[str] = None,
        page_size: int = 100,
        page_token: Optional[str] = None,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """Search for clinical trials by condition.

        Args:
            condition: Medical condition to search for
            status: List of statuses (RECRUITING, NOT_YET_RECRUITING, etc.)
            page_size: Number of results per page (max 1000)
            page_token: Token for pagination

        Returns:
            List of raw trial data
        """
        if status is None:
            status = ["RECRUITING", "NOT_YET_RECRUITING", "ACTIVE_NOT_RECRUITING"]

        params = {
            "query.cond": condition,
            "filter.overallStatus": ",".join(status),
            "pageSize": min(page_size, 1000),
            "format": "json",
            "fields": (
                "NCTId,BriefTitle,OfficialTitle,BriefSummary,DetailedDescription,"
                "Condition,InterventionType,InterventionName,InterventionDescription,"
                "Phase,OverallStatus,EnrollmentCount,EnrollmentType,"
                "StartDate,StartDateType,CompletionDate,CompletionDateType,"
                "LeadSponsorName,LeadSponsorClass,"
                "LocationCountry,LocationCity,LocationFacility,LocationStatus,"
                "CentralContactName,CentralContactEMail,CentralContactPhone,"
                "EligibilityCriteria,MinimumAge,MaximumAge,Sex,HealthyVolunteers,"
                "StudyType,PrimaryOutcomeMeasure,SecondaryOutcomeMeasure"
            )
        }

        if page_token:
            params["pageToken"] = page_token

        response = await self._get(self.BASE_URL, params=params)
        data = response.json()

        return data.get("studies", [])

    async def search_all(
        self,
        condition: str,
        status: List[str] = None,
        max_results: int = 1000
    ) -> List[Dict[str, Any]]:
        """Search with automatic pagination to get all results.

        Args:
            condition: Medical condition to search for
            status: List of statuses to filter by
            max_results: Maximum number of results to return

        Returns:
            List of all matching trials
        """
        all_trials = []
        page_token = None

        while len(all_trials) < max_results:
            params = {
                "query.cond": condition,
                "pageSize": min(100, max_results - len(all_trials)),
                "format": "json"
            }

            if status:
                params["filter.overallStatus"] = ",".join(status)

            if page_token:
                params["pageToken"] = page_token

            response = await self._get(self.BASE_URL, params=params)
            data = response.json()

            studies = data.get("studies", [])
            if not studies:
                break

            all_trials.extend(studies)
            page_token = data.get("nextPageToken")

            if not page_token:
                break

        return all_trials[:max_results]

    async def get_recent_updates(self, days: int = 7) -> List[Dict[str, Any]]:
        """Get trials updated in the last N days."""
        date_from = (datetime.now() - timedelta(days=days)).strftime("%Y-%m-%d")

        all_trials = []
        page_token = None

        while True:
            params = {
                "filter.lastUpdatePostDate": f"[{date_from},]",
                "pageSize": 1000,
                "format": "json"
            }

            if page_token:
                params["pageToken"] = page_token

            response = await self._get(self.BASE_URL, params=params)
            data = response.json()

            studies = data.get("studies", [])
            all_trials.extend(studies)

            page_token = data.get("nextPageToken")
            if not page_token:
                break

        return all_trials

    async def get_by_nct_id(self, nct_id: str) -> Optional[Dict[str, Any]]:
        """Get a specific trial by NCT ID."""
        url = f"{self.BASE_URL}/{nct_id}"
        try:
            response = await self._get(url)
            return response.json()
        except Exception:
            return None

    def normalize(self, raw_trial: Dict[str, Any]) -> Dict[str, Any]:
        """Normalize ClinicalTrials.gov data to common format."""
        protocol = raw_trial.get("protocolSection", {})

        id_module = protocol.get("identificationModule", {})
        status_module = protocol.get("statusModule", {})
        desc_module = protocol.get("descriptionModule", {})
        design_module = protocol.get("designModule", {})
        eligibility_module = protocol.get("eligibilityModule", {})
        contacts_module = protocol.get("contactsLocationsModule", {})
        sponsor_module = protocol.get("sponsorCollaboratorsModule", {})
        conditions_module = protocol.get("conditionsModule", {})
        arms_module = protocol.get("armsInterventionsModule", {})

        # Extract locations
        locations = contacts_module.get("locations", [])
        countries = list(set(loc.get("country", "") for loc in locations if loc.get("country")))
        cities = list(set(loc.get("city", "") for loc in locations if loc.get("city")))
        facilities = [loc.get("facility", "") for loc in locations if loc.get("facility")]

        # Extract contact
        central_contacts = contacts_module.get("centralContacts", [])
        contact = central_contacts[0] if central_contacts else {}

        # Extract interventions
        interventions = arms_module.get("interventions", [])
        intervention_types = list(set(i.get("type", "") for i in interventions if i.get("type")))
        intervention_names = [i.get("name", "") for i in interventions if i.get("name")]

        # Parse dates
        start_date = self._parse_date(status_module.get("startDateStruct", {}))
        completion_date = self._parse_date(status_module.get("completionDateStruct", {}))

        # Get phases
        phases = design_module.get("phases", [])
        phase_str = ", ".join(phases) if phases else None

        nct_id = id_module.get("nctId", "")

        return {
            "nct_id": nct_id,
            "eudract_id": None,
            "who_id": None,
            "isrctn_id": None,

            "title_original": id_module.get("briefTitle") or id_module.get("officialTitle"),
            "summary_original": desc_module.get("briefSummary"),
            "eligibility_original": eligibility_module.get("eligibilityCriteria"),

            "condition": conditions_module.get("conditions", []),
            "intervention_type": intervention_types[0] if intervention_types else None,
            "intervention_name": ", ".join(intervention_names[:3]) if intervention_names else None,

            "phase": phase_str,
            "status": status_module.get("overallStatus"),
            "enrollment_target": design_module.get("enrollmentInfo", {}).get("count"),

            "start_date": start_date,
            "completion_date": completion_date,

            "sponsor": sponsor_module.get("leadSponsor", {}).get("name"),
            "lead_sponsor_type": sponsor_module.get("leadSponsor", {}).get("class"),

            "location_countries": countries,
            "location_cities": cities,
            "location_facilities": facilities,

            "contact_name": contact.get("name"),
            "contact_email": contact.get("email"),
            "contact_phone": contact.get("phone"),

            "age_min": eligibility_module.get("minimumAge"),
            "age_max": eligibility_module.get("maximumAge"),
            "gender": eligibility_module.get("sex"),

            "source_registry": self.source_name,
            "source_url": f"https://clinicaltrials.gov/study/{nct_id}",
            "source_ids": {"nct_id": nct_id},

            "_sources": [self.source_name]
        }

    def _parse_date(self, date_struct: Dict) -> Optional[str]:
        """Parse date from ClinicalTrials.gov date structure."""
        if not date_struct:
            return None

        date_str = date_struct.get("date")
        if not date_str:
            return None

        # Handle various date formats
        for fmt in ["%Y-%m-%d", "%Y-%m", "%B %Y", "%B %d, %Y"]:
            try:
                dt = datetime.strptime(date_str, fmt)
                return dt.strftime("%Y-%m-%d")
            except ValueError:
                continue

        return date_str  # Return as-is if can't parse
